File: kotlin/src/main/kotlin/engine/Grid.kt
package org.minesweeper.engine

import kotlin.random.Random

class Grid(val rows: Int, val cols: Int) {
    constructor(size: Int) : this(size, size)

    val size: Int get() = rows
    val cells: Array<Array<Cell>> = Array(rows) { Array(cols) { Cell() } }
    var flagsPlaced: Int = 0
        private set

    fun placeMines(numberOfMines: Int) {
        var placed = 0
        while (placed < numberOfMines) {
            val row = Random.nextInt(rows)
            val col = Random.nextInt(cols)
            if (!cells[row][col].isMine) {
                cells[row][col].isMine = true
                placed++
            }
        }
        computeAdjacency()
    }

    // First-click safe: avoid 3x3 area around (safeRow, safeCol).
    fun placeMinesSafe(numberOfMines: Int, safeRow: Int, safeCol: Int) {
        val safe = mutableSetOf<Pair<Int, Int>>()
        for (dr in -1..1) for (dc in -1..1) {
            val r = safeRow + dr; val c = safeCol + dc
            if (r in 0 until rows && c in 0 until cols) safe.add(r to c)
        }
        val candidates = mutableListOf<Pair<Int, Int>>()
        for (r in 0 until rows) for (c in 0 until cols)
            if ((r to c) !in safe) candidates.add(r to c)
        candidates.shuffle(Random)
        val n = minOf(numberOfMines, candidates.size)
        for (i in 0 until n) {
            val (r, c) = candidates[i]
            cells[r][c].isMine = true
        }
        computeAdjacency()
    }

    fun placeMinesManually(minePositions: List<Point>) {
        minePositions.forEach { p ->
            if (p.row in 0 until rows && p.col in 0 until cols) cells[p.row][p.col].isMine = true
            else throw IndexOutOfBoundsException("Position out of grid bounds")
        }
        computeAdjacency()
    }

    private fun computeAdjacency() {
        for (r in 0 until rows) for (c in 0 until cols)
            if (!cells[r][c].isMine) cells[r][c].adjacentMines = countAdjacentMines(r, c)
    }

    fun uncoverCell(row: Int, col: Int): Cell {
        if (row !in 0 until rows || col !in 0 until cols)
            throw IndexOutOfBoundsException("Selected cell is out of the grid bounds")
        val cell = cells[row][col]
        if (cell.isRevealed || cell.isFlagged) return cell
        cell.isRevealed = true
        if (cell.isMine) return cell
        cell.adjacentMines = countAdjacentMines(row, col)
        if (cell.adjacentMines == 0) uncoverAdjacentCells(row, col)
        return cell
    }

    fun uncoverAdjacentCells(row: Int, col: Int) {
        for (i in row - 1..row + 1)
            for (j in col - 1..col + 1)
                if (i in 0 until rows && j in 0 until cols && !cells[i][j].isRevealed)
                    uncoverCell(i, j)
    }

    fun countAdjacentMines(row: Int, col: Int): Int {
        var count = 0
        for (i in -1..1) for (j in -1..1) {
            if (i == 0 && j == 0) continue
            val nr = row + i; val nc = col + j
            if (nr in 0 until rows && nc in 0 until cols && cells[nr][nc].isMine) count++
        }
        return count
    }

    fun toggleFlag(row: Int, col: Int) {
        val cell = cells[row][col]
        if (cell.isRevealed) return
        when {
            !cell.isFlagged && !cell.isQuestioned -> { cell.isFlagged = true; flagsPlaced++ }
            cell.isFlagged -> { cell.isFlagged = false; cell.isQuestioned = true; flagsPlaced-- }
            else -> { cell.isQuestioned = false }
        }
    }

    fun chordReveal(row: Int, col: Int): List<Cell> {
        val cell = cells[row][col]
        val revealed = mutableListOf<Cell>()
        if (!cell.isRevealed || cell.adjacentMines == 0) return revealed
        var flagged = 0
        val hidden = mutableListOf<Pair<Int, Int>>()
        for (i in -1..1) for (j in -1..1) {
            if (i == 0 && j == 0) continue
            val nr = row + i; val nc = col + j
            if (nr !in 0 until rows || nc !in 0 until cols) continue
            val n = cells[nr][nc]
            if (n.isFlagged) flagged++
            else if (!n.isRevealed) hidden.add(nr to nc)
        }
        if (flagged != cell.adjacentMines) return revealed
        for ((hr, hc) in hidden) revealed.add(uncoverCell(hr, hc))
        return revealed
    }

    fun revealAllMines() {
        for (r in 0 until rows) for (c in 0 until cols) {
            val cell = cells[r][c]
            if (cell.isMine && !cell.isFlagged) cell.isRevealed = true
            if (!cell.isMine && cell.isFlagged) { cell.wrongFlag = true; cell.isRevealed = true }
        }
    }

    fun uncoverAdjacentCells(point: Point) = uncoverAdjacentCells(point.row, point.col)
    fun uncoverCell(point: Point): Cell = uncoverCell(point.row, point.col)
    fun countAdjacentMines(point: Point): Int = countAdjacentMines(point.row, point.col)
    fun getCell(row: Int, col: Int): Cell = cells[row][col]
    fun getCell(point: Point): Cell = getCell(point.row, point.col)
}

File: csharp/Minesweeper.Engine/Grid.cs
namespace MineSweeper.Engine
{
    public class Grid
    {
        public int Size { get; private set; }
        public int Rows { get; private set; }
        public int Cols { get; private set; }
        public Cell[,] Cells { get; private set; }
        public int FlagsPlaced { get; private set; } = 0;

        public Grid(int size) : this(size, size) { }

        public Grid(int rows, int cols)
        {
            Rows = rows;
            Cols = cols;
            Size = rows; // legacy: rows for square grids
            Cells = new Cell[rows, cols];
            InitializeCells();
        }

        private void InitializeCells()
        {
            for (int i = 0; i < Rows; i++)
                for (int j = 0; j < Cols; j++)
                    Cells[i, j] = new Cell();
        }

        public void PlaceMines(int numberOfMines)
        {
            var rand = new Random();
            int placed = 0;
            while (placed < numberOfMines)
            {
                int row = rand.Next(Rows);
                int col = rand.Next(Cols);
                if (!Cells[row, col].IsMine)
                {
                    Cells[row, col].IsMine = true;
                    placed++;
                }
            }
            ComputeAdjacency();
        }

        // First-click-safe placement: avoid a 3x3 block around (safeRow, safeCol).
        public void PlaceMinesSafe(int numberOfMines, int safeRow, int safeCol)
        {
            var rand = new Random();
            var safe = new HashSet<(int, int)>();
            for (int dr = -1; dr <= 1; dr++)
                for (int dc = -1; dc <= 1; dc++)
                {
                    int rr = safeRow + dr, cc = safeCol + dc;
                    if (rr >= 0 && rr < Rows && cc >= 0 && cc < Cols)
                        safe.Add((rr, cc));
                }
            var candidates = new List<(int, int)>();
            for (int r = 0; r < Rows; r++)
                for (int c = 0; c < Cols; c++)
                    if (!safe.Contains((r, c)))
                        candidates.Add((r, c));
            for (int i = candidates.Count - 1; i > 0; i--)
            {
                int j = rand.Next(i + 1);
                (candidates[i], candidates[j]) = (candidates[j], candidates[i]);
            }
            int n = Math.Min(numberOfMines, candidates.Count);
            for (int i = 0; i < n; i++)
            {
                var (r, c) = candidates[i];
                Cells[r, c].IsMine = true;
            }
            ComputeAdjacency();
        }

        public void PlaceMinesManually(List<Point> minePositions)
        {
            foreach (var position in minePositions)
            {
                int row = position.Row;
                int col = position.Col;
                if (row >= 0 && row < Rows && col >= 0 && col < Cols)
                    Cells[row, col].IsMine = true;
                else
                    throw new ArgumentOutOfRangeException("Position out of grid bounds");
            }
            ComputeAdjacency();
        }

        private void ComputeAdjacency()
        {
            for (int r = 0; r < Rows; r++)
                for (int c = 0; c < Cols; c++)
                    if (!Cells[r, c].IsMine)
                        Cells[r, c].AdjacentMines = CountAdjacentMines(r, c);
        }

        public Cell UncoverCell(int row, int col)
        {
            if (row < 0 || row >= Rows || col < 0 || col >= Cols)
                throw new ArgumentOutOfRangeException("Selected cell is out of the grid bounds.");

            var cell = Cells[row, col];
            if (cell.IsRevealed || cell.IsFlagged) return cell;

            cell.IsRevealed = true;
            if (cell.IsMine) return cell;

            int adjacentMines = CountAdjacentMines(row, col);
            cell.AdjacentMines = adjacentMines;
            if (adjacentMines == 0)
                UncoverAdjacentCells(row, col);
            return cell;
        }

        public void UncoverAdjacentCells(int row, int col)
        {
            for (int i = row - 1; i <= row + 1; i++)
                for (int j = col - 1; j <= col + 1; j++)
                    if (i >= 0 && i < Rows && j >= 0 && j < Cols && !Cells[i, j].IsRevealed)
                        UncoverCell(i, j);
        }

        public int CountAdjacentMines(int row, int col)
        {
            int mineCount = 0;
            for (int i = -1; i <= 1; i++)
                for (int j = -1; j <= 1; j++)
                {
                    if (i == 0 && j == 0) continue;
                    int newRow = row + i, newCol = col + j;
                    if (newRow >= 0 && newRow < Rows && newCol >= 0 && newCol < Cols && Cells[newRow, newCol].IsMine)
                        mineCount++;
                }
            return mineCount;
        }

        public void ToggleFlag(int row, int col)
        {
            var cell = Cells[row, col];
            if (cell.IsRevealed) return;
            if (!cell.IsFlagged && !cell.IsQuestioned)
            {
                cell.IsFlagged = true;
                FlagsPlaced++;
            }
            else if (cell.IsFlagged)
            {
                cell.IsFlagged = false;
                cell.IsQuestioned = true;
                FlagsPlaced--;
            }
            else
            {
                cell.IsQuestioned = false;
            }
        }

        public List<Cell> ChordReveal(int row, int col)
        {
            var cell = Cells[row, col];
            var revealed = new List<Cell>();
            if (!cell.IsRevealed || cell.AdjacentMines == 0) return revealed;

            int flagged = 0;
            var hidden = new List<(int, int)>();
            for (int i = -1; i <= 1; i++)
                for (int j = -1; j <= 1; j++)
                {
                    if (i == 0 && j == 0) continue;
                    int nr = row + i, nc = col + j;
                    if (nr < 0 || nr >= Rows || nc < 0 || nc >= Cols) continue;
                    var n = Cells[nr, nc];
                    if (n.IsFlagged) flagged++;
                    else if (!n.IsRevealed) hidden.Add((nr, nc));
                }
            if (flagged != cell.AdjacentMines) return revealed;
            foreach (var (hr, hc) in hidden)
                revealed.Add(UncoverCell(hr, hc));
            return revealed;
        }

        public void RevealAllMines()
        {
            for (int r = 0; r < Rows; r++)
                for (int c = 0; c < Cols; c++)
                {
                    var cell = Cells[r, c];
                    if (cell.IsMine && !cell.IsFlagged) cell.IsRevealed = true;
                    if (!cell.IsMine && cell.IsFlagged)
                    {
                        cell.WrongFlag = true;
                        cell.IsRevealed = true;
                    }
                }
        }

        public void UncoverAdjacentCells(Point point) => UncoverAdjacentCells(point.Row, point.Col);
        public Cell UncoverCell(Point point) => UncoverCell(point.Row, point.Col);
        public int CountAdjacentMines(Point point) => CountAdjacentMines(point.Row, point.Col);
        public Cell GetCell(int row, int col) => Cells[row, col];
        public Cell GetCell(Point cell) => GetCell(cell.Row, cell.Col);
    }
}
